ssl/ssl_key_share.cc

```diff
@@ -290,13 +293,197 @@ class CECPQ2KeyShare : public SSLKeyShare {
+    if (!EVP_PQ_KEM_encapsulate(pq_kem_ctx_)) {
+      *out_alert = SSL_AD_DECODE_ERROR;
+      return false;
+    }
+
+    if (!CBB_add_u16(out_public_key, pq_kem_ctx_->kem->ciphertext_length) ||
+        !CBB_add_bytes(out_public_key, pq_kem_ctx_->ciphertext,pq_kem_ctx_->kem->ciphertext_length) ||
+        !out_secret ||
+        !out_secret->Init(ec_secret.size() + pq_kem_ctx_->kem->shared_secret_length)) {
+      return false;
+    }
+
+    OPENSSL_memcpy(out_secret->data(), ec_secret.data(), ec_secret.size());
+    OPENSSL_memcpy(out_secret->data() + ec_secret.size(), pq_kem_ctx_->shared_secret,
+                   pq_kem_ctx_->kem->shared_secret_length);
+
+    EVP_PQ_KEM_CTX_free(pq_kem_ctx_);
+
+    return true;
+  }
+
+  bool Finish(Array<uint8_t> *out_secret, uint8_t *out_alert,
+              Span<const uint8_t> peer_key) override {
+    *out_alert = SSL_AD_INTERNAL_ERROR;
+    assert(pq_kem_ctx_);
+    assert(ec_key_share_);
+
+    uint16_t ec_public_key_length;
+    Array<uint8_t> ec_secret;
+
+    if (!get_ec_public_key_length(&ec_public_key_length, ec_nid_)) {
+      return false;
+    }
+
+    // Verify all lengths
+    if (peer_key.size() != ec_public_key_length + pq_kem_ctx_->kem->ciphertext_length + (2 * sizeof(uint16_t))) {
+      *out_alert = SSL_AD_DECODE_ERROR;
+      return false;
+    }
+    uint16_t received_ec_size, received_pq_size;
+    CBS ec_cbs, pq_cbs;
+    CBS_init(&ec_cbs, peer_key.data(), sizeof(uint16_t));
+    CBS_init(&pq_cbs, peer_key.data() + sizeof(uint16_t) + ec_public_key_length, sizeof(uint16_t));
+    if (!CBS_get_u16(&ec_cbs, &received_ec_size) ||
+        received_ec_size != ec_public_key_length ||
+        !CBS_get_u16(&pq_cbs, &received_pq_size) ||
+        received_pq_size != pq_kem_ctx_->kem->ciphertext_length) {
+      *out_alert = SSL_AD_DECODE_ERROR;
+      return false;
+    }
+
+    Span<const uint8_t> ec_peer_key = peer_key.subspan(sizeof(uint16_t), ec_public_key_length);
+    if (!ec_key_share_->Finish(&ec_secret, out_alert, ec_peer_key)) {
+      *out_alert = SSL_AD_DECODE_ERROR;
+      return false;
+    }
+
+    OPENSSL_memcpy(pq_kem_ctx_->ciphertext, peer_key.data() + ec_public_key_length + (2 * sizeof(uint16_t)),
+                   pq_kem_ctx_->kem->ciphertext_length);
+    if (!EVP_PQ_KEM_decapsulate(pq_kem_ctx_)) {
+      *out_alert = SSL_AD_DECODE_ERROR;
+      return false;
+    }
+
+    if (!out_secret ||
+        !out_secret->Init(ec_secret.size() + pq_kem_ctx_->kem->shared_secret_length)) {
+      return false;
+    }
+
+    OPENSSL_memcpy(out_secret->data(), ec_secret.data(), ec_secret.size());
+    OPENSSL_memcpy(out_secret->data() + ec_secret.size(), pq_kem_ctx_->shared_secret,
+                   pq_kem_ctx_->kem->shared_secret_length);
+
+    EVP_PQ_KEM_CTX_free(pq_kem_ctx_);
+
+    return true;
+  }
+
+ private:
+  int ec_nid_;
+  int pq_nid_;
+  uint16_t group_id_;
+  UniquePtr<SSLKeyShare> ec_key_share_;
+  EVP_PQ_KEM_CTX *pq_kem_ctx_;
```

Comment:
do you think it's worth it to implement `PQKeyShare`class similarly to `SSLKeyShare` and `CECPQ2KeyShare` and then in `PQHybridKeyShare` just use `PQKeyShare` and `SSLKeyShare`?